agents/postgres/pgstatmonitor/pgstatmonitor.go

```diff
@@ -0,0 +1,302 @@
+			}
+
+			m.changes <- agents.Change{MetricsBucket: buckets}
+		}
+	}
+}
+
+func (m *PGStatMonitorQAN) getNewBuckets(ctx context.Context, periodStart time.Time, periodLengthSecs uint32) ([]*agentpb.MetricsBucket, error) {
+	current, prev, err := m.monitorCache.getStatMonitorExtended(ctx, m.q)
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := makeBuckets(current, prev, m.disableQueryExamples, m.pgsmNormalizedQuery, m.l)
+	startS := uint32(periodStart.Unix())
+	m.l.Debugf("Made %d buckets out of %d stat monitor in %s+%d interval.",
+		len(buckets), len(current), periodStart.Format("15:04:05"), periodLengthSecs)
+
+	// merge prev and current in cache
+	m.monitorCache.refresh(current)
+	m.l.Debugf("statMonitorCache: %s", m.monitorCache.stats())
+
+	// add agent_id and timestamps
+	for i, b := range buckets {
+		b.Common.AgentId = m.agentID
+		b.Common.PeriodStartUnixSecs = startS
+		b.Common.PeriodLengthSecs = periodLengthSecs
+
+		buckets[i] = b
+	}
+
+	return buckets, nil
+}
+
+// makeBuckets uses current state of pg_stat_monitor table and accumulated previous state
+// to make metrics buckets.
+//
+// makeBuckets is a pure function for easier testing.
+func makeBuckets(current, prev map[string]*pgStatMonitorExtended, disableQueryExamples, pgsmNormalizedQuery bool, l *logrus.Entry) []*agentpb.MetricsBucket {
+	res := make([]*agentpb.MetricsBucket, 0, len(current))
+
+	for queryID, currentPSS := range current {
+		prevPSS := prev[queryID]
+		if prevPSS == nil {
+			prevPSS = new(pgStatMonitorExtended)
+		}
+		count := float32(currentPSS.Calls - prevPSS.Calls)
+		switch {
+		case count == 0:
+			// Another way how this is possible is if pg_stat_monitor was truncated,
+			// and then the same number of queries were made.
+			// Currently, we can't differentiate between those situations.
+			l.Tracef("Skipped due to the same number of queries: %s.", currentPSS)
+			continue
+		case count < 0 || (time.Now().Sub(currentPSS.BucketStartTime).Seconds() > 60):
+			l.Debugf("Truncate detected. Treating as a new query: %s.", currentPSS)
+			prevPSS = new(pgStatMonitorExtended)
+			count = float32(currentPSS.Calls)
+		case prevPSS.Calls == 0:
+			l.Debugf("New query: %s.", currentPSS)
+		default:
+			l.Debugf("Normal query: %s.", currentPSS)
+		}
+
+		fingerprint := currentPSS.Query
+		example := ""
+		if !pgsmNormalizedQuery {
+			fingerprint = query.Fingerprint(currentPSS.Query)
+			example = currentPSS.Query
+		}
+
+		mb := &agentpb.MetricsBucket{
+			Common: &agentpb.MetricsBucket_Common{
+				Fingerprint: fingerprint,
+				Database:    currentPSS.Database,
+				Tables:      currentPSS.TablesNames,
+				Username:    currentPSS.Username,
+				Queryid:     currentPSS.QueryID,
+				NumQueries:  count,
+				AgentType:   inventorypb.AgentType_QAN_POSTGRESQL_PGSTATMONITOR_AGENT,
+			},
+			Postgresql: new(agentpb.MetricsBucket_PostgreSQL),
+		}
+
+		if example != "" && !disableQueryExamples {
+			example, truncated := truncate.Query(example)
```

Comment:
let's truncate it here https://github.com/percona/pmm-agent/pull/183/files#diff-c7bb32cfdcfd2dcb1693bf29def551dcR124 and use `IsQueryTruncated` from `pgStatMonitorExtended`